system info NullPointerException

In a brand-new installation based on openhabian (on an rpi 3), the OH2 system info binding throws a null pointer exception on startup and does not update any channels.

Config: (no cfg file)
`Thing systeminfo:computer:raspberrypi [interval_high=15, interval_medium=120] 

Group gSystem
Number cpuMemTotal     "Memory total [%.0f MBytes]"      (gSystem) { channel="systeminfo:computer:raspberrypi:memory#total" }
Number cpuMemUsed      "Memory used  [%.0f MBytes]"      (gSystem) { channel="systeminfo:computer:raspberrypi:memory#used" }
Number cpuMemAvailable "Memory available [%.0f MBytes]"  (gSystem) { channel="systeminfo:computer:raspberrypi:memory#available" }
Number cpuMemPercent   "Memory available [%.0f %%]"      (gSystem) { channel="systeminfo:computer:raspberrypi:memory#availablePercent" }

Number diskTotal       "Disk total [%.0f MBytes]"        (gSystem) { channel="systeminfo:computer:raspberrypi:storage#total" }
Number diskPercent     "Disk available [%.0f %%]"        (gSystem) { channel="systeminfo:computer:raspberrypi:storage#availablePercent" }
Number diskAvailable   "Disk available [%.0f MBytes]"    (gSystem) { channel="systeminfo:computer:raspberrypi:storage#available" }
Number cpuLoad         "CPU load [%.0f %%]"              (gSystem) { channel="systeminfo:computer:raspberrypi:cpu#load" }
`

Log:

`12:50:27.296 [INFO ] [me.event.ThingStatusInfoChangedEvent] - 'systeminfo:computer:raspberrypi' changed from UNINITIALIZED to INITIALIZING
12:50:27.452 [ERROR] [platform.linux.LinuxCentralProcessor] - Couldn't find physical processor count. Assuming 1.
12:50:27.842 [INFO ] [me.event.ThingStatusInfoChangedEvent] - 'systeminfo:computer:raspberrypi' changed from INITIALIZING to ONLINE
12:50:28.351 [ERROR] [ome.core.thing.internal.ThingManager] - Exception occured while calling thing updated at ThingHandler 'org.openhab.binding.systeminfo.handler.SysteminfoHandler@12d0bb1: java.lang.NullPointerException
java.util.concurrent.ExecutionException: java.lang.NullPointerException
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)[:1.8.0_111]
	at java.util.concurrent.FutureTask.get(FutureTask.java:206)[:1.8.0_111]
	at org.eclipse.smarthome.core.common.SafeMethodCaller.callAsynchronous(SafeMethodCaller.java:186)[96:org.eclipse.smarthome.core:0.9.0.201611181452]
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:79)[96:org.eclipse.smarthome.core:0.9.0.201611181452]
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:63)[96:org.eclipse.smarthome.core:0.9.0.201611181452]
	at org.eclipse.smarthome.core.thing.internal.ThingManager.thingUpdated(ThingManager.java:498)[103:org.eclipse.smarthome.core.thing:0.9.0.201611181452]
	at org.eclipse.smarthome.core.thing.internal.ThingRegistryImpl.notifyTrackers(ThingRegistryImpl.java:226)[103:org.eclipse.smarthome.core.thing:0.9.0.201611181452]
	at org.eclipse.smarthome.core.thing.internal.ThingRegistryImpl.notifyListenersAboutUpdatedElement(ThingRegistryImpl.java:149)[103:org.eclipse.smarthome.core.thing:0.9.0.201611181452]
	at org.eclipse.smarthome.core.thing.internal.ThingRegistryImpl.notifyListenersAboutUpdatedElement(ThingRegistryImpl.java:1)[103:org.eclipse.smarthome.core.thing:0.9.0.201611181452]
	at org.eclipse.smarthome.core.common.registry.AbstractRegistry.updated(AbstractRegistry.java:167)[96:org.eclipse.smarthome.core:0.9.0.201611181452]
	at org.eclipse.smarthome.core.common.registry.AbstractProvider.notifyListeners(AbstractProvider.java:57)[96:org.eclipse.smarthome.core:0.9.0.201611181452]
	at org.eclipse.smarthome.core.common.registry.AbstractProvider.notifyListenersAboutUpdatedElement(AbstractProvider.java:82)[96:org.eclipse.smarthome.core:0.9.0.201611181452]
	at org.eclipse.smarthome.model.thing.internal.GenericThingProvider.access$3(GenericThingProvider.java:1)[131:org.eclipse.smarthome.model.thing:0.9.0.201611181452]
	at org.eclipse.smarthome.model.thing.internal.GenericThingProvider$1$2.apply(GenericThingProvider.java:985)[131:org.eclipse.smarthome.model.thing:0.9.0.201611181452]
	at org.eclipse.smarthome.model.thing.internal.GenericThingProvider$1$2.apply(GenericThingProvider.java:1)[131:org.eclipse.smarthome.model.thing:0.9.0.201611181452]
	at org.eclipse.xtext.xbase.lib.IteratorExtensions.forEach(IteratorExtensions.java:363)[144:org.eclipse.xtext.xbase.lib:2.9.2]
	at org.eclipse.xtext.xbase.lib.IterableExtensions.forEach(IterableExtensions.java:333)[144:org.eclipse.xtext.xbase.lib:2.9.2]
	at org.eclipse.smarthome.model.thing.internal.GenericThingProvider$1.run(GenericThingProvider.java:994)[131:org.eclipse.smarthome.model.thing:0.9.0.201611181452]
	at java.lang.Thread.run(Thread.java:745)[:1.8.0_111]
Caused by: java.lang.NullPointerException
	at org.openhab.binding.systeminfo.handler.SysteminfoHandler.handleChannelConfigurationChange(SysteminfoHandler.java:568)[194:org.openhab.binding.systeminfo:2.0.0.201611181914]
	at org.openhab.binding.systeminfo.handler.SysteminfoHandler.thingUpdated(SysteminfoHandler.java:547)[194:org.openhab.binding.systeminfo:2.0.0.201611181914]
	at org.eclipse.smarthome.core.thing.internal.ThingManager$6.call(ThingManager.java:502)[103:org.eclipse.smarthome.core.thing:0.9.0.201611181452]
	at org.eclipse.smarthome.core.thing.internal.ThingManager$6.call(ThingManager.java:1)[103:org.eclipse.smarthome.core.thing:0.9.0.201611181452]
	at org.eclipse.smarthome.core.common.SafeMethodCaller$CallableWrapper.call(SafeMethodCaller.java:177)[96:org.eclipse.smarthome.core:0.9.0.201611181452]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_111]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)[:1.8.0_111]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)[:1.8.0_111]
	... 1 more
`

--- a/addons/binding/org.openhab.binding.systeminfo.test/src/main/groovy/org/openhab/binding/systeminfo/test/SysteminfoOSGiTest.groovy
+++ b/addons/binding/org.openhab.binding.systeminfo.test/src/main/groovy/org/openhab/binding/systeminfo/test/SysteminfoOSGiTest.groovy
@@ -830,6 +830,39 @@ class SysteminfoOSGiTest extends OSGiTest{
         testItemStateIsUpdated(acceptedItemType, DEFAULT_TEST_ITEM_NAME, DEFAULT_CHANNEL_TEST_PRIORITY)
     }
 
+    @Test
+    public void 'test thing handles channel priority change' () {
+        def priorityKey = "priority"
+        def pidKey = "pid"
+        def initialPriority = DEFAULT_CHANNEL_TEST_PRIORITY // Evaluates to High
+        def newPriority = "Low"
+
+        String acceptedItemType = "Number"
+        initializeThingWithChannel(DEFAULT_TEST_CHANNEL_ID,acceptedItemType)
+
+
+        Channel channel = systemInfoThing.getChannel(DEFAULT_TEST_CHANNEL_ID)
+
+        waitForAssert {
+            assertThat "The initial priority of channel ${channel.getUID()} is not as expected.", channel.getConfiguration().get(priorityKey), is (equalTo(initialPriority))
+            assertThat systemInfoThing.getHandler().highPriorityChannels.contains(channel.getUID()), is (true)
+        }
+
+        //Change the priority of a channel, keep the pid
+        Configuration updatedConfig = new Configuration()
+        updatedConfig.put(priorityKey, newPriority)
+        updatedConfig.put(pidKey, channel.getConfiguration().get(pidKey))
+        Channel updatedChannel = new Channel (channel.getUID(),channel.getChannelTypeUID(),channel.getAcceptedItemType(),channel.getKind(),updatedConfig,new HashSet(),new HashMap(),null,null)
+        Thing updatedThing = ThingBuilder.create(systemInfoThing.getThingTypeUID(),systemInfoThing.getUID()).withConfiguration(systemInfoThing.getConfiguration()).withChannel(updatedChannel).build();
+
+        systemInfoThing.getHandler().thingUpdated(updatedThing)
+
+        waitForAssert {
+            assertThat "The prority of the channel was not updated: ", channel.getConfiguration().get(priorityKey), is (equalTo(newPriority))
+            assertThat systemInfoThing.getHandler().lowPriorityChannels.contains(channel.getUID()), is (true)
+        }
+    }
+
     @After
     public void tearDown () {
 

--- a/addons/binding/org.openhab.binding.systeminfo/src/main/java/org/openhab/binding/systeminfo/handler/SysteminfoHandler.java
+++ b/addons/binding/org.openhab.binding.systeminfo/src/main/java/org/openhab/binding/systeminfo/handler/SysteminfoHandler.java
@@ -11,7 +11,6 @@ package org.openhab.binding.systeminfo.handler;
 import static org.openhab.binding.systeminfo.SysteminfoBindingConstants.*;
 
 import java.math.BigDecimal;
-import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Iterator;
 import java.util.List;
@@ -86,26 +85,18 @@ public class SysteminfoHandler extends BaseThingHandler {
     ScheduledFuture<?> highPriorityTasks;
     ScheduledFuture<?> mediumPriorityTasks;
 
-    private Configuration currentThingConfiguration;
-    private Map<ChannelUID, Configuration> currentChannelConfiguration;
-
     private Logger logger = LoggerFactory.getLogger(SysteminfoHandler.class);
 
     public SysteminfoHandler(Thing thing) {
         super(thing);
-        currentThingConfiguration = new Configuration();
-        currentChannelConfiguration = new HashMap<ChannelUID, Configuration>();
     }
 
     @Override
     public void initialize() {
         logger.debug("Start initializing!");
 
-        updateThingConfiguration();
-        updateChannelConfiguration();
-
         if (instantiateSysteminfoLibrary() && isConfigurationValid() && updateProperties()) {
-            groupChannelsByPriority(currentChannelConfiguration);
+            groupChannelsByPriority();
             scheduleUpdates();
             logger.debug("Thing is successfully initialized!");
             updateStatus(ThingStatus.ONLINE);
@@ -130,8 +121,9 @@ public class SysteminfoHandler extends BaseThingHandler {
     private boolean isConfigurationValid() {
         logger.debug("Start reading Thing configuration.");
         try {
-            refreshIntervalMediumPriority = (BigDecimal) currentThingConfiguration.get(MEDIUM_PRIORITY_REFRESH_TIME);
-            refreshIntervalHighPriority = (BigDecimal) currentThingConfiguration.get(HIGH_PRIORITY_REFRESH_TIME);
+            refreshIntervalMediumPriority = (BigDecimal) this.thing.getConfiguration()
+                    .get(MEDIUM_PRIORITY_REFRESH_TIME);
+            refreshIntervalHighPriority = (BigDecimal) this.thing.getConfiguration().get(HIGH_PRIORITY_REFRESH_TIME);
 
             if (refreshIntervalHighPriority.intValue() <= 0 || refreshIntervalMediumPriority.intValue() <= 0) {
                 throw new IllegalArgumentException("Refresh time must be positive number!");
@@ -159,31 +151,33 @@ public class SysteminfoHandler extends BaseThingHandler {
             logger.debug("Properties updated!");
             return true;
         } catch (Exception e) {
-            logger.debug("Can not get system properties!. Please try to restart the binding.", e);
+            logger.debug("Can not get system properties! Please try to restart the binding.", e);
             return false;
         }
 
     }
 
-    private void groupChannelsByPriority(Map<ChannelUID, Configuration> channelConfiguration) {
-        Set<ChannelUID> channelUIDs = channelConfiguration.keySet();
+    private void groupChannelsByPriority() {
+        logger.trace("Grouping channels by priority.");
+        List<Channel> channels = this.thing.getChannels();
 
-        for (ChannelUID channelUID : channelUIDs) {
-            Configuration properties = channelConfiguration.get(channelUID);
+        for (Channel channel : channels) {
+            Configuration properties = channel.getConfiguration();
             String priority = (String) properties.get(PRIOIRITY_PARAM);
             if (priority == null) {
-                logger.debug("Channel with id {} will not be updated. The channel has no priority set !", channelUID);
+                logger.debug("Channel with id {} will not be updated. The channel has no priority set !",
+                        channel.getUID());
                 break;
             }
             switch (priority) {
                 case "High":
-                    highPriorityChannels.add(channelUID);
+                    highPriorityChannels.add(channel.getUID());
                     break;
                 case "Medium":
-                    mediumPriorityChannels.add(channelUID);
+                    mediumPriorityChannels.add(channel.getUID());
                     break;
                 case "Low":
-                    lowPriorityChannels.add(channelUID);
+                    lowPriorityChannels.add(channel.getUID());
                     break;
                 default:
                     logger.error("Invalid priority configuration parameter. Channel will not be updated !");
@@ -474,7 +468,7 @@ public class SysteminfoHandler extends BaseThingHandler {
     private int getPID(ChannelUID channelUID) {
         int pid = 0;
         try {
-            Configuration channelProperties = currentChannelConfiguration.get(channelUID);
+            Configuration channelProperties = this.thing.getChannel(channelUID.getId()).getConfiguration();
             BigDecimal pidValue = (BigDecimal) channelProperties.get(PID_PARAM);
             if (pidValue.intValue() < 0) {
                 throw new IllegalArgumentException("Invalid value for Process Identifier.");
@@ -504,23 +498,9 @@ public class SysteminfoHandler extends BaseThingHandler {
         }
     }
 
-    private void updateThingConfiguration() {
-        currentThingConfiguration = thing.getConfiguration();
-        logger.debug("Thing configuration updated!");
-    }
-
-    private void updateChannelConfiguration() {
-        List<Channel> channels = thing.getChannels();
-        for (Channel channel : channels) {
-            currentChannelConfiguration.put(channel.getUID(), channel.getConfiguration());
-        }
-
-        logger.debug("Channel configuration updated!");
-    }
-
-    private boolean isConfigurationKeyChanged(Configuration currentConifg, Configuration newConfig, String key) {
-        if (currentConifg != null) {
-            Object currentValue = currentConifg.get(key);
+    private boolean isConfigurationKeyChanged(Configuration currentConfig, Configuration newConfig, String key) {
+        if (currentConfig != null && newConfig != null) {
+            Object currentValue = currentConfig.get(key);
             Object newValue = newConfig.get(key);
 
             if (currentValue == null) {
@@ -534,17 +514,26 @@ public class SysteminfoHandler extends BaseThingHandler {
 
     @Override
     public void thingUpdated(Thing thing) {
+        logger.trace("About to update thing.");
         boolean isChannelConfigChanged = false;
         List<Channel> channels = thing.getChannels();
+
         for (Channel channel : channels) {
             ChannelUID channelUID = channel.getUID();
             Configuration newChannelConfig = channel.getConfiguration();
-            Configuration currentChannelConfig = currentChannelConfiguration.get(channelUID);
+            Channel oldChannel = this.thing.getChannel(channelUID.getId());
+
+            if (oldChannel == null) {
+                logger.warn("Channel with UID : {} can not be updated, as it can not be found !",
+                        channelUID.getAsString());
+                continue;
+            }
+            Configuration currentChannelConfig = oldChannel.getConfiguration();
 
             if (isConfigurationKeyChanged(currentChannelConfig, newChannelConfig, PRIOIRITY_PARAM)) {
                 isChannelConfigChanged = true;
 
-                handleChannelConfigurationChange(channelUID, newChannelConfig, PRIOIRITY_PARAM);
+                handleChannelConfigurationChange(oldChannel, newChannelConfig, PRIOIRITY_PARAM);
 
                 String newPriority = (String) newChannelConfig.get(PRIOIRITY_PARAM);
                 changeChannelPriority(channelUID, newPriority);
@@ -552,25 +541,25 @@ public class SysteminfoHandler extends BaseThingHandler {
 
             if (isConfigurationKeyChanged(currentChannelConfig, newChannelConfig, PID_PARAM)) {
                 isChannelConfigChanged = true;
-                handleChannelConfigurationChange(channelUID, newChannelConfig, PID_PARAM);
+                handleChannelConfigurationChange(oldChannel, newChannelConfig, PID_PARAM);
             }
         }
 
-        if (isInitialized() && isChannelConfigChanged) {
-            updateChannelConfiguration();
-        } else {
+        if (!(isInitialized() && isChannelConfigChanged)) {
             super.thingUpdated(thing);
         }
-    };
+    }
 
-    private void handleChannelConfigurationChange(ChannelUID channelUID, Configuration newConfig, String parameter) {
-        Channel channel = this.thing.getChannel(channelUID.getId());
+    private void handleChannelConfigurationChange(Channel channel, Configuration newConfig, String parameter) {
         Configuration configuration = channel.getConfiguration();
-        configuration.setProperties(newConfig.getProperties());
+        Object oldValue = configuration.get(parameter);
 
-        Object value = newConfig.get(parameter);
-        logger.debug("Channel with UID : {} has changed its {} to : {}", channelUID.getAsString(), parameter, value);
-        publishDataForChannel(channelUID);
+        configuration.put(parameter, newConfig.get(parameter));
+
+        Object newValue = newConfig.get(parameter);
+        logger.debug("Channel with UID : {} has changed its {} from {} to {}", channel.getUID(), parameter, oldValue,
+                newValue);
+        publishDataForChannel(channel.getUID());
     }
 
     private void stopScheduledUpdates() {

